ddbToEsDlqLambda/index.ts

```diff
@@ -0,0 +1,151 @@
+        .invoke({
+            FunctionName: DDB_TO_ES_LAMBDA_NAME,
+            InvocationType: INVOCATION_TYPE,
+            Payload: JSON.stringify(records),
+        })
+        .promise()
+        .catch((e) => {
+            throw new Error(`Failed to invoke DdbToEsLambda: ${JSON.stringify(e)}`);
+        });
+
+    // StatusCode 200 only means successful invocation.
+    if (resp.StatusCode >= 400) {
+        throw new Error(`Failed to invoke DdbToEsLambda: ${JSON.stringify(resp.Payload)}`);
+    }
+
+    // When errors occur during execution, StatusCode is 200, but FunctionError is present.
+    if (resp.FunctionError) {
+        throw new Error(`Failed to invoke DdbToEsLambda: ${JSON.stringify(resp.Payload)}`);
+    }
+
+    logger.debug('Invoked DdbToEsLambda with records', JSON.stringify(records));
+};
+
+const getRecordsFromDdbStream = async (message) => {
+    const resp = await dynamodbstreams
+        .getShardIterator({
+            ShardId: message.DDBStreamBatchInfo.shardId,
+            ShardIteratorType: 'AT_SEQUENCE_NUMBER',
+            StreamArn: message.DDBStreamBatchInfo.streamArn,
+            SequenceNumber: message.DDBStreamBatchInfo.startSequenceNumber,
+        })
+        .promise()
+        .catch((e) => {
+            throw new Error(`Failed to get shard iterator: ${JSON.stringify(e)}`);
+        });
+
+    const records = await dynamodbstreams
+        .getRecords({
+            ShardIterator: resp.ShardIterator,
+        })
+        .promise()
+        .catch((e) => {
+            throw new Error(`Failed to get records: ${JSON.stringify(e)}`);
+        });
+
+    logger.debug('Fetched records', JSON.stringify(records));
+    return records;
+};
+
+/**
+ * This Lambda function will read message from the DDbToEs DLQ, fetch records from the DynamoDB stream
+ * referenced in the queued message, and invoke the DDBToEs Lambda function providing the records as input.
+ *
+ * Note:
+ *   1. Message retention period is 14 days. After retention period, messages would be deleted automatically.
+ *   2. Record retention period is 24 hours. After retention period, records would be deleted automatically.
+ *   3. Message visibility timeout must be greater than the total time of processing the message,
+ *      otherwise the message would be received by this Lambda function again. Too long timeout would
+ *      make the messages invisible for long time, and it appears there are no messages in DLQ.
+ */
+exports.handler = async (event) => {
+    let numMessagesToProcess = event.maxNumberOfMessages
+        ? event.maxNumberOfMessages
+        : Number(MAX_NUMBER_OF_MESSAGES_TO_PROCESS);
+
+    if (typeof numMessagesToProcess !== 'number' || numMessagesToProcess < 1) {
+        throw new Error(`invalid maxNumberOfMessages: ${numMessagesToProcess}`);
+    }
+
+    while (numMessagesToProcess > 0) {
+        // If SQS error is thrown, then stop processing messages.
+        /* eslint-disable no-await-in-loop */
```

Comment:
Do we want a try catch here so a single batch failure does not fail the whole process?

Reply:
The thought was if we encounter an error communicating with SQS, either retrieving or deleting messages off of the queue, then we might as well not try anymore. Return a failed result to the caller, and let them try again later.

Reply:
got it, yeah if it's just AWS receiveMessage, invokeLambda and deleteMessage we shouldn't have any head of line blocking on logical errors.